bigbang/ingress/abstract.py

```diff
@@ -0,0 +1,773 @@
+        containing units of year, month, and week-of-month which can have the following
+        example elements:
+            - years: (1992, 2010), [2000, 2008], 2021
+            - months: ["January", "July"], "November"
+            - weeks: (1, 4), [1, 5], 2
+
+        Parameters
+        ----------
+        times : A list containing information of the period for each
+            group of mboxMessage.
+        urls : Corresponding URLs of each group of mboxMessage of which the
+            period info is contained in `times`.
+        filtr : Containing info on what should be filtered.
+
+        Returns
+        -------
+        Indices of to the elements in `times`/`ursl`.
+        """
+        if isinstance(filtr, tuple):
+            # filter year or week in range
+            cond = lambda x: (np.min(filtr) <= x <= np.max(filtr))
+        if isinstance(filtr, list):
+            # filter in year, week, or month in list
+            cond = lambda x: x in filtr
+        if isinstance(filtr, int):
+            # filter specific year or week
+            cond = lambda x: x == filtr
+        if isinstance(filtr, str):
+            # filter specific month
+            cond = lambda x: x == filtr
+        return [idx for idx, time in enumerate(times) if cond(time)]
+
+    @abstractmethod
+    def get_name_from_url(url: str) -> str:
+        pass
+
+    def to_dict(self, include_body: bool = True) -> Dict[str, List[str]]:
+        """
+        Parameters
+        ----------
+        include_body : A boolean that indicates whether the message body should
+            be included or not.
+
+        Returns
+        -------
+        A Dictionary with the first key layer being the header field names and
+        the "body" key. Each value field is a list containing the respective
+        header field contents arranged by the order as they were scraped from
+        the web. This format makes the conversion to a pandas.DataFrame easier.
+        """
+        return ListIO.to_dict(self.messages, include_body)
+
+    def to_pandas_dataframe(self, include_body: bool = True) -> pd.DataFrame:
+        """
+        Parameters
+        ----------
+        include_body : A boolean that indicates whether the message body should
+            be included or not.
+
+        Returns
+        -------
+        Converts the mailing list into a pandas.DataFrame object in which each
+        row represents an Email.
+        """
+        return ListIO.to_pandas_dataframe(self.messages, include_body)
+
+    def to_mbox(self, dir_out: str, filename: Optional[str] = None):
+        """Safe mailing list to .mbox files."""
+        if filename is None:
+            ListIO.to_mbox(self.messages, dir_out, self.name)
+        else:
+            ListIO.to_mbox(self.messages, dir_out, filename)
+
+
+class AbstractArchive(ABC):
+    """
+    This class handles the scraping of a public mailing list archive.
```

Comment:
Again -- what is the difference between a list and an archive?

I believe that in some notebooks, Archives including email from multiple lists are built and saved to disk.
But these muilt-archives are not scraped directly.

This old Archive design may be something to scrap and not build so much around.

But I wonder how you have used this class in your ListServ work, and how it differs from you use of the List classes.

Reply:
> Again -- what is the difference between a list and an archive?
Find an extended explanation below under the headline: "Difference between list and archive"